# Functions Assignment 1

1. Write a Python function to find the maximum of three numbers. 

In [6]:
def max(a,b,c):
    if b<=a>=c:
        print(f"{a} or a is greater number")
    elif a<=b>=c:
        print(f"{b} or b is greater number")
    else:
        print(f"{c} or c is greater number")

In [7]:
max(13,14,12)

14 or b is greater number


2. Write a Python function to sum all the numbers in a list. 
Sample List : (8, 2, 3, 0, 7) 
Expected Output : 20 

In [1]:
def sum_of_numbers(numbers):
    return sum(numbers)
numbers_list = [1, 2, 3, 4, 5]
total_sum = sum_of_numbers(numbers_list)
print(f"The sum of the numbers in the list is: {total_sum}")

The sum of the numbers in the list is: 15


3. Write a Python function to multiply all the numbers in a list. 
Sample List : (8, 2, 3, -1, 7) 
Expected Output : -336 

In [13]:
def multiply_numbers(numbers):
    result = 1
    for number in numbers:
        result *= number
    return result
numbers_list = [1, 2, 3, 4, 5]
total_product = multiply_numbers(numbers_list)
print(f"The product of the numbers in the list is: {total_product}")


The product of the numbers in the list is: 120


4. Write a Python program to reverse a string. 
Sample String : "1234abcd" 
Expected Output : "dcba4321" 

In [15]:
def reverse_string(s):
    return s[::-1]
# Example usage
original_string = "1234abcd"
reversed_string = reverse_string(original_string)
print(f"The reversed string is: {reversed_string}")


The reversed string is: dcba4321


5. Write a Python function to calculate the factorial of a number (a non-negative 
integer). The function accepts the number as an argument. 

In [4]:
def factorial(n):
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers.")
    elif n == 0:
        return 1
    else:
        return n * factorial(n-1)

In [5]:
number = 5
result = factorial(number)
print(f"The factorial of {number} is {result}")

The factorial of 5 is 120


6. Write a Python function to check whether a number falls within a given range. 

In [13]:
def is_within_range(number, lower_bound, upper_bound):
    return lower_bound <= number <= upper_bound

num = 10
lower = 15
upper = 20

if is_within_range(num, lower, upper):
  print(f"{num} is within the range [{lower}, {upper}]")
else:
  print(f"{num} is not within the range [{lower}, {upper}]")

10 is not within the range [15, 20]


7. Write a Python function that accepts a string and counts the number of upper 
and lower case letters. 
Sample String : 'The quick Brow Fox' 
Expected Output : 
No. of Upper case characters : 3 
No. of Lower case Characters : 12 

In [14]:
def count_upper_lower(string):
  upper_count = 0
  lower_count = 0

  for char in string:
    if char.isupper():
      upper_count += 1
    elif char.islower():
      lower_count += 1

  return upper_count, lower_count

In [15]:
string = 'The quick Brow Fox'
upper, lower = count_upper_lower(string)

print(f"No. of Upper case characters : {upper}")
print(f"No. of Lower case Characters : {lower}")

No. of Upper case characters : 3
No. of Lower case Characters : 12


8. Write a Python function that takes a list and returns a new list with distinct 
elements from the first list. 
Sample List : [1,2,3,3,3,3,4,5] 
Unique List : [1, 2, 3, 4, 5] 

In [18]:
def unique_elements(lst):
  return list(set(lst))

In [19]:
sample_list = [1, 2, 3, 3, 3, 3, 4, 5]
unique_list = unique_elements(sample_list)
print("Sample List:", sample_list)
print("Unique List:", unique_list)

Sample List: [1, 2, 3, 3, 3, 3, 4, 5]
Unique List: [1, 2, 3, 4, 5]


9. Write a Python function that takes a number as a parameter and checks 
whether the number is prime or not. 
Note : A prime number (or a prime) is a natural number greater than 1 and that 
has no positive divisors other than 1 and itself. 

In [27]:
def is_prime(number):
    if number <= 1:
        return False
    for i in range(2, int(number**0.5) + 1):
        if number % i == 0:
              return False
    return True

In [52]:
num = 223
if is_prime(num):
  print(f"{num} is a prime number.")
else:
  print(f"{num} is not a prime number.")

223 is a prime number.


10. Write a Python program to print the even numbers from a given list. 
Sample List : [1, 2, 3, 4, 5, 6, 7, 8, 9]
Expected Result : [2, 4, 6, 8] 

In [36]:
def even_numbers(lst):
    even_numbers = []
    for num in lst:
        if num % 2 == 0:
            even_numbers.append(num)
    return even_numbers

In [37]:
sample_list = [1, 2, 3, 4, 5, 6, 7, 8, 9]
even_numbers_list = even_numbers(sample_list)
print("Sample List:", sample_list)
print("Even Numbers:", even_numbers_list)  # Output: Even Numbers: [2, 4, 6, 8]

Sample List: [1, 2, 3, 4, 5, 6, 7, 8, 9]
Even Numbers: [2, 4, 6, 8]


11. Write a Python function to check whether a number is "Perfect" or not.
A perfect number is a positive integer that is equal to the sum of its proper positive divisors, that is, the sum of its 
positive divisors excluding the number itself (also known as its aliquot sum). Equivalently, a perfect number is a number that is half the sum of all of its positive divisors (including itself). 
Example : The first perfect number is 6, because 1, 2, and 3 are its proper positive 
divisors, and 1 + 2 + 3 = 6. Equivalently, the number 6 is equal to half the sum of 
all its positive divisors: ( 1 + 2 + 3 + 6 ) / 2 = 6. The next perfect number is 28 = 1 + 
2 + 4 + 7 + 14. This is followed by the perfect numbers 496 and 8128. 

In [39]:
def is_perfect_number(number):
    if number <= 0:
        return False
    sum_of_divisors = 0
    for i in range(1, number):
        if number % i == 0:
            sum_of_divisors += i
    return sum_of_divisors == number

In [44]:
num = 12
if is_perfect_number(num):
  print(f"{num} is a perfect number.")
else:
  print(f"{num} is not a perfect number.")

12 is not a perfect number.


12. Write a Python function that checks whether a passed string is a palindrome 
or not. 
Note: A palindrome is a word, phrase, or sequence that reads the same backward 
as forward, e.g., madam or nurses run. 

In [46]:
def is_palindrome(string):
    clean_string = string.replace(" ", "").lower()
    reversed_string = clean_string[::-1]
    return clean_string == reversed_string

In [47]:
string1 = "racecar"
string2 = "A man a plan a canal Panama"

if is_palindrome(string1):
  print(f"{string1} is a palindrome.")
else:
  print(f"{string1} is not a palindrome.")

if is_palindrome(string2):
  print(f"{string2} is a palindrome.")
else:
  print(f"{string2} is not a palindrome.")

racecar is a palindrome.
A man a plan a canal Panama is a palindrome.


# FUNCTIONS ASSIGNMENT 2

Task1: 
You need to write a function in python that return all such numbers which are divisible by 7 but are not a multiple of 5, 
between 2000 and 3200 (both included). 
Return the obtained numbers in a comma-separated sequence on a single line.

In [1]:
def multiple_of_7():
    list=[]
    for i in range(2000,3201):
        if i%7==0:
            list.append(str(i))
    return ",".join(list)

In [2]:
multiple_of_7()

'2002,2009,2016,2023,2030,2037,2044,2051,2058,2065,2072,2079,2086,2093,2100,2107,2114,2121,2128,2135,2142,2149,2156,2163,2170,2177,2184,2191,2198,2205,2212,2219,2226,2233,2240,2247,2254,2261,2268,2275,2282,2289,2296,2303,2310,2317,2324,2331,2338,2345,2352,2359,2366,2373,2380,2387,2394,2401,2408,2415,2422,2429,2436,2443,2450,2457,2464,2471,2478,2485,2492,2499,2506,2513,2520,2527,2534,2541,2548,2555,2562,2569,2576,2583,2590,2597,2604,2611,2618,2625,2632,2639,2646,2653,2660,2667,2674,2681,2688,2695,2702,2709,2716,2723,2730,2737,2744,2751,2758,2765,2772,2779,2786,2793,2800,2807,2814,2821,2828,2835,2842,2849,2856,2863,2870,2877,2884,2891,2898,2905,2912,2919,2926,2933,2940,2947,2954,2961,2968,2975,2982,2989,2996,3003,3010,3017,3024,3031,3038,3045,3052,3059,3066,3073,3080,3087,3094,3101,3108,3115,3122,3129,3136,3143,3150,3157,3164,3171,3178,3185,3192,3199'

Task2:
You need code a function that calculates and returns the value according to the given formula: 
P = [(2 * A * B)/C]**1/2 
Following are the fixed values of A and B: 
A is 50. B is 30. 
The values of the literal C should be taken as console input to your program in a comma-separated sequence. 
Example Let us assume the following comma separated input sequence is given to the program: 
100,150,180 
The output of the program should be: 
18,22,24 

In [5]:
import math
def calculate_values(C):
    A = 50
    B = 30
    return [int(math.sqrt((2 * A * B) / c)) for c in C]
# Example usage
C_input = input("Enter the values of C as comma-separated sequence: ")
C_values = [int(c) for c in C_input.split(',')]
output = calculate_values(C_values)
print(", ".join(map(str, output)))

Enter the values of C as comma-separated sequence:  1


54


Task3: 
You need to write a function that takes a comma separated sequence of words as 
input and prints the words in a comma-separated sequence after sorting them 
alphabetically. 
Suppose the following input is supplied to the program: 
without,hello,bag,world 
Then, the output should be: 
bag,hello,without,world 

In [6]:
def sort_words():
    words = input("Enter a comma-separated sequence of words: ")
    word_list = words.split(',')
    sorted_word_list = sorted(word_list)
    sorted_words = ','.join(sorted_word_list)
    print(sorted_words)

In [7]:
sort_words()

Enter a comma-separated sequence of words:  this,my,bag,is


bag,is,my,this


Task4: 
You need to write a program that takes sequence of lines as input and prints the 
lines after making all characters in the sentence capitalized. 
Suppose the following input is supplied to the program: 
Hello world 
Practice makes perfect 
Then, the output should be: 
HELLO WORLD 
PRACTICE MAKES PERFECT 

In [8]:
def capitalize_lines():
    lines = []
    while True:
        line = input()
        if not line:
            break
        lines.append(line.upper())
    
    for line in lines:
        print(line)

print("Enter your lines (press Enter twice to stop):")
capitalize_lines()

Task5: 
You need to write a function that counts the number of vowels in a given 
sentence as input from console. 
Suppose the following input is supplied to the program: 
Hello world 
Practice makes perfect 
Then, the output should be: 
a appeared 2 times 
e appeared 5 times 
i appeared 1 time 
o appeared 2 times 
u appeared 0 time

In [12]:
def count_vowels(sentence):
    vowels = 'aeiou'
    vowel_count = {v: 0 for v in vowels}
    sentence = sentence.lower()

    for char in sentence:
        if char in vowels:
            vowel_count[char] += 1
    for vowel in vowels:
        count = vowel_count[vowel]
        if count == 1:
            print(f"{vowel} appeared {count} time")
        else:
            print(f"{vowel} appeared {count} times")

In [13]:
input_sentence = input("Enter a sentence: ")
count_vowels(input_sentence)

Enter a sentence:  hello world


a appeared 0 times
e appeared 1 time
i appeared 0 times
o appeared 2 times
u appeared 0 times


Task6: 
You need write a function that traces and makes a list of all such numbers 
from1000 to 3000 in which all the digits are even numbers. 

In [14]:
def even_no():
    list=[]
    for i in range(1000,3001):
        if i%2==0:
            list.append(str(i))
    return ",".join(list)

In [15]:
even_no()

'1000,1002,1004,1006,1008,1010,1012,1014,1016,1018,1020,1022,1024,1026,1028,1030,1032,1034,1036,1038,1040,1042,1044,1046,1048,1050,1052,1054,1056,1058,1060,1062,1064,1066,1068,1070,1072,1074,1076,1078,1080,1082,1084,1086,1088,1090,1092,1094,1096,1098,1100,1102,1104,1106,1108,1110,1112,1114,1116,1118,1120,1122,1124,1126,1128,1130,1132,1134,1136,1138,1140,1142,1144,1146,1148,1150,1152,1154,1156,1158,1160,1162,1164,1166,1168,1170,1172,1174,1176,1178,1180,1182,1184,1186,1188,1190,1192,1194,1196,1198,1200,1202,1204,1206,1208,1210,1212,1214,1216,1218,1220,1222,1224,1226,1228,1230,1232,1234,1236,1238,1240,1242,1244,1246,1248,1250,1252,1254,1256,1258,1260,1262,1264,1266,1268,1270,1272,1274,1276,1278,1280,1282,1284,1286,1288,1290,1292,1294,1296,1298,1300,1302,1304,1306,1308,1310,1312,1314,1316,1318,1320,1322,1324,1326,1328,1330,1332,1334,1336,1338,1340,1342,1344,1346,1348,1350,1352,1354,1356,1358,1360,1362,1364,1366,1368,1370,1372,1374,1376,1378,1380,1382,1384,1386,1388,1390,1392,1394,1396,1398

Task7: 
You need to write a code which accepts a sequence of comma separated 4 digit 
binary numbers as its input and then check whether they are divisible by 5 or not. 
The numbers that are divisible by 5 are to be printed in a comma separated 
sequence. 
Example: 
0100,0011,1010,1001 
Then the output should be: 
1010 

In [17]:
def divisible_by_five(binary_numbers):
    divisible_numbers = []
    for number in binary_numbers.split(','):
        decimal_number = int(number, 2)  # Convert binary to decimal
        if decimal_number % 5 == 0:
            divisible_numbers.append(number)
    return ','.join(divisible_numbers)

In [18]:
input_string = input("Enter binary numbers (comma-separated): ")
result = divisible_by_five(input_string)
print("Divisible by 5:", result)

Enter binary numbers (comma-separated):  1111,1100,1001,1110


Divisible by 5: 1111


Task8: 
Write a program that accepts a sentence and calculate the number of letters and digits. 
Suppose the following input is supplied to the program: 
hello world! 123 
Then, the output should be: 
LETTERS 10 
DIGITS 3

In [19]:
def count_letters_and_digits(sentence):
    letters = 0
    digits = 0
    for char in sentence:
        if char.isalpha():
            letters += 1
        elif char.isdigit():
            digits += 1
    return letters, digits

In [20]:
sentence = input("Enter a sentence: ")
letters, digits = count_letters_and_digits(sentence)
print(f"LETTERS {letters}")
print(f"DIGITS {digits}")

Enter a sentence:  hello world! 123


LETTERS 10
DIGITS 3
